Create graceful-delegate-shutdown-process.md

kat-enos opened the pull request:
# Harness Developer Pull Request
Thanks for helping us make the Developer Hub better. The PR will be looked at
by the maintainers. 

## What Type of PR is This?

- [ x] Issue
- [ ] Feature
- [ ] Maintenance/Chore

If tied to an Issue, list the Issue(s) here:

* *Issue(s)* https://harness.atlassian.net/browse/DOC-1996

## House Keeping
Some items to keep track of. Screen shots of changes are optional but would help the maintainers review quicker. 

- [ x] Tested Locally
- [ ] *Optional* Screen Shoot.

bot-gitexp-user commented:
Preview environment: https://ci.pr.harness.io/pr-575

michaelcretzman commented on docs/platform/2_Delegates/get-started-with-delegates/graceful-delegate-shutdown-process.md:
```diff
@@ -0,0 +1,29 @@
+---
+title: Graceful delegate shutdown
+description: This document explains how delegate shutdown works based on delegate type and environment.
```
```suggestion
description: Learn how delegate shutdown works based on delegate type and environment.
```

michaelcretzman replied:
just suggesting a shorter description since the cards in our doc site are so small.

kat-enos replied:
I think if I change it I'll stick in the term graceful since this document does not explain much more than that. But will wait to see what an SME says about the below changes before I controvert your approval.

michaelcretzman commented on docs/platform/2_Delegates/get-started-with-delegates/graceful-delegate-shutdown-process.md:
```diff
@@ -0,0 +1,29 @@
+# sidebar_position: 2
+---
+
+Harness Delegate is implemented to shut down gracefully. The process of graceful shutdown is applied as follows:
+
+- Delegate receives an instruction to quit.
+- A grace period begins during which the delegate:
+  + Stops accepting new tasks.
+  + Works to complete running tasks.
+- The grace period ends.
+- Delegates that have not quit are force-terminated.
+- Incomplete tasks are discarded.
```
```suggestion
Harness Delegate shuts down gracefully. The shutdown process follows these steps:

1. Delegate receives an instruction to quit.
2. A grace period begins during which the delegate:
  + Stops accepting new tasks.
  + Works to complete running tasks.
3. The grace period ends.
4. Delegates that have not quit are force-terminated.
5. Incomplete tasks are discarded.
```

michaelcretzman replied:
seems like this is a sequence so an ordered list is more useful.

kat-enos replied:
Always bulleted list when you're just describing a process. At least that's the way I've been taught. Numbers are for procedures with discrete steps.

michaelcretzman commented on docs/platform/2_Delegates/get-started-with-delegates/graceful-delegate-shutdown-process.md:
```diff
@@ -0,0 +1,29 @@
+| **Delegate environment** | **Trigger** |
+| Kubernetes | Pod termination, eviction, or user-initiated scaling |
+| Docker | `docker kill` command |
+| Shell | `./stop.sh` instruction |
```
we have a shell script delegate now? We did in FG, but I didn't know it was in NG.

kat-enos replied:
Yeah, I can't really say. This was given to me and, as you know, I'm still learning. But I'm having a technical reviewer look at it and the PM will take a second look too. I think, actually, that this information was from the issue and that that ticket was opened by a technical lead.

michaelcretzman commented on docs/platform/2_Delegates/get-started-with-delegates/graceful-delegate-shutdown-process.md:
```diff
@@ -0,0 +1,29 @@
+The event that initiates the graceful shutdown depends on delegate type.
+
+| **Delegate environment** | **Trigger** |
+| Kubernetes | Pod termination, eviction, or user-initiated scaling |
```
Sure about this?

The K8s Delegate is a deployment object. 

When a Kubernetes pod is terminated in a Deployment object, the Deployment's replica count is decremented and a new pod is created to replace the terminated pod. The new pod will have the same configuration as the previous pod and continue to run the same containers and associated processes. This is because Deployments are designed to ensure that the desired number of replicas are running at all times, and to provide automatic self-healing of pods in case of failures or terminations.

I would image you would need to terminate the Deployment to initiate shutdown. Like we cover here:

https://developer.harness.io/docs/platform/delegates/manage-delegates/delete-a-delegate/

kat-enos replied:
I'm familiar with how replica sets work but from what I see here, I think pod termination could still trigger a graceful delegate shutdown process. Raghu or another SME from the team will review this. Thank you.

michaelcretzman reviewed (dismissed):
approving with the assumption that you will consider my suggestions.

pratmit commented on docs/platform/2_Delegates/get-started-with-delegates/graceful-delegate-shutdown-process.md:
```diff
@@ -0,0 +1,29 @@
+- Delegates that have not quit are force-terminated.
+- Incomplete tasks are discarded.
+
+The length of the grace period is configurable and depends on whether the delegate is immutable or legacy. 
```
remove the context to legacy, just say delegate with the immutable image

kat-enos replied:
@pratmit This was changed to "The length of the grace period is configurable."

pratmit reviewed (approved):
Assuming the changes are accepted

kat-enos commented on docs/platform/2_Delegates/get-started-with-delegates/graceful-delegate-shutdown-process.md:
```diff
@@ -0,0 +1,29 @@
+- Delegates that have not quit are force-terminated.
+- Incomplete tasks are discarded.
+
+The length of the grace period is configurable and depends on whether the delegate is immutable or legacy. 
```
```suggestion
The length of the grace period is configurable. 
```

kat-enos commented:
@pratmit I don't find any comments from Raghu in response to some issues Michael raised. I responded to Michael's comments by saying that pod deletion could be a trigger for the graceful shutdown process. You approved this document after I made that response so it sounds like we agree on that.

I implemented your changes.  You asked for the elimination of references to legacy in an intro sentence and I removed it. But after that, there is a table that has two lines -- one for immutable and one for legacy as follows:


| **Delegate type** | **Grace period** | **Default interval** |
| :-- | :--: | :--: |
| Immutable | Yes | 10 minutes |
| Legacy | No | 30 seconds |

Do you want me to strike the row with the legacy information? If so, I will put the immutable information into sentence form and merge the content.

Thanks.

bot-gitexp-user commented:
Preview environment: https://hdh.pr.harness.io/pr-575

pratmit commented on docs/platform/2_Delegates/get-started-with-delegates/graceful-delegate-shutdown-process.md:
```diff
@@ -0,0 +1,30 @@
+
+| **Delegate type** | **Grace period** | **Default interval** |
+| :-- | :--: | :--: |
+| Immutable | Yes | 10 minutes |
```
Immutable Image

kat-enos replied:
```suggestion
| Immutable image | Yes | 10 minutes |
```

pratmit commented on docs/platform/2_Delegates/get-started-with-delegates/graceful-delegate-shutdown-process.md:
```diff
@@ -0,0 +1,30 @@
+| **Delegate type** | **Grace period** | **Default interval** |
+| :-- | :--: | :--: |
+| Immutable | Yes | 10 minutes |
+| Legacy | No | 30 seconds |
```
Legacy Image

kat-enos replied:
```suggestion
| Legacy image | No | 30 seconds |
```

kat-enos commented:
@pratmit, your latest changes should be integrated. Let me know if you want me to merge it. Thanks.

pratmit reviewed (approved):
LGTM

bot-gitexp-user commented:
Preview environment: https://hdh.pr.harness.io/pr-575

kat-enos commented:
Thanks @pratmit. Merging.
